Update PermissionsPolicy to match latest spec

szewai opened the pull request:
#### 061e4cd3970fe03212bc57c761cb3d5e4088421f
<pre>
Update PermissionsPolicy to match latest spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=273982">https://bugs.webkit.org/show_bug.cgi?id=273982</a>
<a href="https://rdar.apple.com/127848428">rdar://127848428</a>

Reviewed by Youenn Fablet.

This patch implements some algorithms and does some refactoring/renaming in PermissionPolicy based on latest version of
spec.

* Source/WebCore/Modules/applepay/PaymentSession.cpp:
(WebCore::PaymentSession::canCreateSession):
* Source/WebCore/Modules/audiosession/DOMAudioSession.cpp:
(WebCore::DOMAudioSession::setType):
(WebCore::DOMAudioSession::type const):
(WebCore::DOMAudioSession::state const):
(WebCore::DOMAudioSession::scheduleStateChangeEvent):
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
* Source/WebCore/Modules/geolocation/Geolocation.cpp:
(WebCore::Geolocation::shouldBlockGeolocationRequests):
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::checkCameraAccess):
(WebCore::checkMicrophoneAccess):
(WebCore::checkSpeakerAccess):
(WebCore::MediaDevices::listenForDeviceChanges):
* Source/WebCore/Modules/mediastream/UserMediaController.cpp:
(WebCore::UserMediaController::logEnumerateDevicesDenial):
* Source/WebCore/Modules/mediastream/UserMediaRequest.cpp:
(WebCore::UserMediaRequest::start):
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::isAllowedByPermissionsPolicy):
* Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
(WebCore::WakeLock::request):
* Source/WebCore/Modules/speech/SpeechRecognition.cpp:
(WebCore::SpeechRecognition::startRecognition):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
(WebCore::WebXRSystem::isSessionSupported):
(WebCore::WebXRSystem::isFeaturePermitted const):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::isFullscreenEnabled const):
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::permissionsPolicy const):
* Source/WebCore/html/PermissionsPolicy.cpp:
(WebCore::toFeatureNameForLogging):
(WebCore::readFeatureIdentifier):
(WebCore::splitOnAsciiWhiteSpace):
(WebCore::isPermissionsPolicyAllowedByDocumentAndAllOwners):
(WebCore::PermissionsPolicy::Allowlist::matches const):
(WebCore::PermissionsPolicy::parseAllowlist):
(WebCore::PermissionsPolicy::parsePolicyDirective):
(WebCore::PermissionsPolicy::declaredOrigin const):
(WebCore::PermissionsPolicy::PermissionsPolicy):
(WebCore::PermissionsPolicy::allows const):
(WebCore::policyTypeName): Deleted.
(WebCore::isAllowedByPermissionsPolicy): Deleted.
(WebCore::processOriginItem): Deleted.
(WebCore::updateList): Deleted.
(WebCore::PermissionsPolicy::parse): Deleted.
* Source/WebCore/html/PermissionsPolicy.h:
(WebCore::PermissionsPolicy::defaultPolicy):
(WebCore::PermissionsPolicy::parse):
(WebCore::PermissionsPolicy::Allowlist::Allowlist):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::makeXRCompatible):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::isAllowedToUseDeviceMotion const):
(WebCore::LocalDOMWindow::isAllowedToUseDeviceOrientation const):
* Source/WebCore/page/Navigator.cpp:
(WebCore::validateWebSharePolicy):
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::createRequest):

Canonical link: <a href="https://commits.webkit.org/278786@main">https://commits.webkit.org/278786@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9bb49979c7e50bb336ff70c6bdacdcc4bd0f5137

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51537 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30847 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3889 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54804 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2230 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53840 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1910 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41958 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28485 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44448 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23084 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25791 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1729 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47750 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1794 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56396 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26656 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1668 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49356 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27896 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44517 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48546 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11281 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28789 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27631 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -39,62 +39,171 @@ namespace WebCore {
 }
 
-bool isPermissionsPolicyAllowedByDocumentAndAllOwners(PermissionsPolicy::Type type, const Document& document, LogPermissionsPolicyFailure logFailure)
+static std::pair<PermissionsPolicy::Feature, StringView> splitFeatureName(StringView value)
```
Add https://w3c.github.io/webappsec-permissions-policy/#serialized-policy-directive link?
Maybe we should reuse the terminology as well.

szewai replied:
ok, will add link and rename it `readFeatureIdentifier`

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -39,62 +39,171 @@ namespace WebCore {
+    }
+
+    if (remainingValue.startsWith(":"_s))
+        remainingValue = remainingValue.substring(1);
```
Maybe add a FIXME to remove this rule?
If we can do that, we could potentially further align and check for whitespace as a way to split the feature name.

szewai replied:
Will add.

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -123,315 +232,148 @@ bool isPermissionsPolicyAllowedByDocumentAndAllOwners(PermissionsPolicy::Type ty
-        return false;
-    case PermissionsPolicy::AllowRule::Type::All:
+    if (HashSet<SecurityOriginData>* origins = std::get_if<HashSet<SecurityOriginData>>(&m_origins))
+        origins->add(origin);
```
This is a bit strange, auto* might be slightly better,.
The if might not be needed in practice?
I would tend to build a HashSet<SecurityOriginData>, then once filled, create an AllowList from it.

szewai replied:
I don't quite understand this comment; are you suggesting `Allowlist` should own a `HashSet` instead of `std::variant`?

youennf replied:
I would remove addOrigin altogether and add an `explicit PermissionsPolicy::Allowlist(HashSet&&)`

szewai replied:
ok will update

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -123,315 +232,148 @@ bool isPermissionsPolicyAllowedByDocumentAndAllOwners(PermissionsPolicy::Type ty
-        }
+    PermissionsPolicy::PolicyDirective result;
+    for (auto item : value.split(';')) {
+        // FIXME: Based on spec, we should use splitOnAsciiWhiteSpace to extract feature name,
```
Ah I see there is a FIXME, let's file a bugzilla, especially if other browsers are not doing this.

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -123,315 +232,148 @@ bool isPermissionsPolicyAllowedByDocumentAndAllOwners(PermissionsPolicy::Type ty
+        // FIXME: Based on spec, we should use splitOnAsciiWhiteSpace to extract feature name,
+        // but some tests would fail as they use ':' as delimiter.
+        auto [feature, remainingItem] = splitFeatureName(item);
+        if (feature == Feature::Invalid)
```
I guess we could remove Feature::Invalid and return an optional. Either is fine.

youennf commented on Source/WebCore/html/PermissionsPolicy.h:
```diff
@@ -63,45 +64,36 @@ class PermissionsPolicy {
+    using PolicyDirective = HashMap<Feature, Allowlist, IntHash<Feature>, WTF::StrongEnumHashTraits<Feature>>;
+    PolicyDirective parsePolicyDirective(StringView, const SecurityOriginData& containerOrigin, const SecurityOriginData& targetOrigin, bool useStarAsDefaultAllowlistValue);
+
+    PolicyDirective m_effectivePolicy;
```
This is more a PermissionsPolicy, a PolicyDirective would be a single entry, not a HashMap.
I would tend to rename PolicyDirective to PolicyDirectives and m_effectivePolicy to m_policies.

szewai replied:
The [spec](https://w3c.github.io/webappsec-permissions-policy/#policy-directives) says "A policy directive is an [ordered map](https://infra.spec.whatwg.org/#ordered-map), mapping [policy-controlled features](https://w3c.github.io/webappsec-permissions-policy/#policy-controlled-feature) to corresponding [allowlists](https://w3c.github.io/webappsec-permissions-policy/#allowlist) of origins."
(I will add a link comment above `using PolicyDirective...`)

youennf replied:
Ah I see, then it is fine.
I was looking at https://w3c.github.io/webappsec-permissions-policy/#serialized-policy-directive where it is a single entry.
I guess the spec is not entirely consistent. It is fine then.

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -123,315 +232,148 @@ bool isPermissionsPolicyAllowedByDocumentAndAllOwners(PermissionsPolicy::Type ty
+    if (iter == m_effectivePolicy.end())
+        return false;
+
+    return iter->value.matches(origin);
```
single ternary line?

youennf reviewed (approved):
LGTM

szewai committed:
Update PermissionsPolicy to match latest spec
https://bugs.webkit.org/show_bug.cgi?id=273982
rdar://127848428

Reviewed by Youenn Fablet.

This patch implements some algorithms and does some refactoring/renaming in PermissionPolicy based on latest version of
spec.

* Source/WebCore/Modules/applepay/PaymentSession.cpp:
(WebCore::PaymentSession::canCreateSession):
* Source/WebCore/Modules/audiosession/DOMAudioSession.cpp:
(WebCore::DOMAudioSession::setType):
(WebCore::DOMAudioSession::type const):
(WebCore::DOMAudioSession::state const):
(WebCore::DOMAudioSession::scheduleStateChangeEvent):
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
* Source/WebCore/Modules/geolocation/Geolocation.cpp:
(WebCore::Geolocation::shouldBlockGeolocationRequests):
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::checkCameraAccess):
(WebCore::checkMicrophoneAccess):
(WebCore::checkSpeakerAccess):
(WebCore::MediaDevices::listenForDeviceChanges):
* Source/WebCore/Modules/mediastream/UserMediaController.cpp:
(WebCore::UserMediaController::logEnumerateDevicesDenial):
* Source/WebCore/Modules/mediastream/UserMediaRequest.cpp:
(WebCore::UserMediaRequest::start):
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::isAllowedByPermissionsPolicy):
* Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
(WebCore::WakeLock::request):
* Source/WebCore/Modules/speech/SpeechRecognition.cpp:
(WebCore::SpeechRecognition::startRecognition):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
(WebCore::WebXRSystem::isSessionSupported):
(WebCore::WebXRSystem::isFeaturePermitted const):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::isFullscreenEnabled const):
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::permissionsPolicy const):
* Source/WebCore/html/PermissionsPolicy.cpp:
(WebCore::toFeatureNameForLogging):
(WebCore::readFeatureIdentifier):
(WebCore::splitOnAsciiWhiteSpace):
(WebCore::isPermissionsPolicyAllowedByDocumentAndAllOwners):
(WebCore::PermissionsPolicy::Allowlist::matches const):
(WebCore::PermissionsPolicy::parseAllowlist):
(WebCore::PermissionsPolicy::parsePolicyDirective):
(WebCore::PermissionsPolicy::declaredOrigin const):
(WebCore::PermissionsPolicy::PermissionsPolicy):
(WebCore::PermissionsPolicy::allows const):
(WebCore::policyTypeName): Deleted.
(WebCore::isAllowedByPermissionsPolicy): Deleted.
(WebCore::processOriginItem): Deleted.
(WebCore::updateList): Deleted.
(WebCore::PermissionsPolicy::parse): Deleted.
* Source/WebCore/html/PermissionsPolicy.h:
(WebCore::PermissionsPolicy::defaultPolicy):
(WebCore::PermissionsPolicy::parse):
(WebCore::PermissionsPolicy::Allowlist::Allowlist):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::makeXRCompatible):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::isAllowedToUseDeviceMotion const):
(WebCore::LocalDOMWindow::isAllowedToUseDeviceOrientation const):
* Source/WebCore/page/Navigator.cpp:
(WebCore::validateWebSharePolicy):
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::createRequest):

Canonical link: https://commits.webkit.org/278786@main
